### ISSUE TITLE
[0.9.0 staging-1594] Law may exist without attaching to any court

### ISSUE THREAD

**Issue by SlayksWood** [StrangeLoopGames/EcoIssues#16601]
First of all this is a bug, but We met this bug 3 times for all period. First time it was near 3 week ago after big changes in civics and it was in migration playtest. Second and third time was yesterday and today. But we can't reproduce it.
But the worst tning that I can't delete this law with dev tier. And this law works.
It happens after revise law. So the simplest steps:
- create any law 1.
- /civics winelection
- revise law 1. It should rename it to law 2.
- /civics winelection. Law 1 will be removed, law 2 should be move to law cell, but sometimes (
I would say very very rarely) this law will removed from court, but it still exists.
So I provide save files with such law:
https://drive.google.com/file/d/1XlwyvwMtxwVs2kiyvdTDYVhnqvqQjzxh/view?usp=sharing
- just load this world and join to it.
- open government, legislation tab. You can see 2 laws:
![image](https://user-images.githubusercontent.com/45708377/84125057-af135800-aa44-11ea-93e1-3b08a15a0463.png)
- law 4 is bugged law. In tooltip you can see that it will be dissabled after 20 min, but this happen only after restart and I can easy fix it. Just broke room:
![image](https://user-images.githubusercontent.com/45708377/84125240-f0a40300-aa44-11ea-9b51-22183b67a047.png)
- and repair it (or maybe you just need to wait some time):
![image](https://user-images.githubusercontent.com/45708377/84125348-16c9a300-aa45-11ea-99f3-124cc113fd01.png)
- So next try to see why this law works. This law is contained on court. But if you check this court then you can see that information is wrong:
![image](https://user-images.githubusercontent.com/45708377/84125586-690ac400-aa45-11ea-80b9-d4a105f4531e.png)
![image](https://user-images.githubusercontent.com/45708377/84125617-745def80-aa45-11ea-969e-271c3584e145.png)

We will continue to investigate the bug with the possibility of this situation, but in any case need to fix existing such law. This law should be disabled. Or we need to make admin command to delete this law.

**Comment by johnkslg** [comment]
It can be removed with /objects remove laws <law id>
you can remove most anything like that. Ill look at causes and auto fixes.

**Comment by johnkslg** [comment]
Made init process detect orphaned civicobjects/contracts/work parties and readd them:
![image](https://user-images.githubusercontent.com/3536496/84209540-48edfb80-aa6b-11ea-8575-68225f6c120b.png)
It adds them to the first open slot, or creates a new slot (exceeding the limit of 3) if none are open.

![image](https://user-images.githubusercontent.com/3536496/84209778-dfbab800-aa6b-11ea-824d-44529750a287.png)

**Comment by SlayksWood** [comment]
hm i didn't find this command, I assumed it should be finded in /civics.

**Comment by SlayksWood** [comment]
Still an issue for elected titles (at least):
![image](https://user-images.githubusercontent.com/45708377/85013232-eda7c180-b16c-11ea-8112-d07cb6edf1b4.png)
And it wasnot added after server restart.
Save files:
https://drive.google.com/file/d/1ZAiG_KUpctnfM632uIPVXx6OSZ8vkKud/view?usp=sharing
- I have this glitch. And then submit 3 new elected titles. After that I restart server.

**Comment by johnkslg** [comment]
Cant repro, Loaded that linked world and the title was on the requested spot just fine.
Can you give me a world where it repros?
Also, during my stream while I was making laws they were disappearing from the court as they finished elections, I then tried locally and couldnt make it happen. Maybe try in that branch I was in? Can see what I was doing in the stream recording to put together some repro steps.

**Comment by MishkaFur** [comment]
> 
> 
> Cant repro, Loaded that linked world and the title was on the requested spot just fine.
> Can you give me a world where it repros?
> Also, during my stream while I was making laws they were disappearing from the court as they finished elections, I then tried locally and couldnt make it happen. Maybe try in that branch I was in? Can see what I was doing in the stream recording to put together some repro steps.

We are having this issue on playtest but its seemingly at random, some laws with a few revisions are okay but others are not.

Economy court according to the law highlight has "Tax Band v2 2" but it does not, also Road funding does not appear on any court

**Comment by WeaselDog** [comment]
This issue just appeared on white-tiger playtest when force passing a law amendment with /fin. Doing the same steps in local did not replicate the issue.

**Comment by johnkslg** [comment]
@beastyn  @SlayksWood need a solid repro case

**Comment by IcedForge** [comment]
Add law, use Modify law to revise a condition use /fin 16 ( id of the law in question ) and it vanished from its slot Current White Tiger
using 0.9.0.0 staging-1702
![Screenshot_280](https://user-images.githubusercontent.com/24981362/89194094-31347e80-d5a7-11ea-8238-746311ae277f.jpg)
As you can tell by the notifications it passed and after that it vanished, ive done this several times now through these 2 days and it seems consistent with the issue of modifying a draft, this happens in not just laws but also district zonings as well.

**Comment by IcedForge** [comment]
![Screenshot_268](https://user-images.githubusercontent.com/24981362/89195985-dea89180-d5a9-11ea-9292-9350741b88c5.jpg)
Example of when i did it 3 times on the same slot yesterday before it was restarted, on server restart it had 2 extra slots to account for the "re add to hosted court"

**Comment by SlayksWood** [comment]
@johnkslg 
There are no steps for this. It just might happen. I can't find a suitable case to reproduce it. It doesn't happen very often, but when it happens it isn't very pleasant.

**Comment by WeaselDog** [comment]
This should be solved by this: https://github.com/StrangeLoopGames/Eco/pull/6605

**Comment by beastyn** [comment]
New video and log #17786

Its in 1728, I hope it will gone after 1730

**Comment by beastyn** [comment]
@MishkaFur @IcedForge 

You've got the new update which should include a fix for the issue.
Can I ask you to post here any related news?

**Comment by IcedForge** [comment]
Ill have a go at it in a fresh world to see if this is still an issue then,thank you beastyn

**Comment by IcedForge** [comment]
I have so far been unsuccessful at recreating the issues - But had a report on discord that the issue is still happening.
![Screenshot_108](https://user-images.githubusercontent.com/24981362/91182082-ff1bc580-e6e9-11ea-99c2-0c9567c3995c.jpg)

Asked them to contact me and post their logs etc here and comment when their admin for the server comes online.

**Comment by beastyn** [comment]
@IcedForge thanks a lot, Logs and also save would be really helpful. I will make QA pass for civics too these days, hope to catch it.

**Comment by beastyn** [comment]
@johnkslg I was able to repro it today (1757) just with patient and time (I also added 300 test clients and launch a second client for some kind of simulation. But I'm sure it's not necessary). Create an elected title - propose - /fin election number - revise - propose - /fin election number - repeat from revise,
I think you can try to repro it with command. Just launch this chain in auto mode till table lose an object.
Anyway, it's just Kirill said - it won't create a new version of an object after it will be placed into removed state (you even can see this removed mark on a table's cell). It will remove it from a table, but object itself will still exist. For me it was about 50 trys

**Comment by smokeykiller51** [comment]
If the laws disappearing are rare then we got really really unlucky on Panda. I think part of the problem was that we had multiple people interacting with gov building in one non-active capital then he would go back to where he had the active government building and propose more laws. We had one title which passed and election Presidente. Then after being elected the title was assigned to me. There was another proposed title el presidente. Which was only in the proposed state initially. After the server restarted Presidente disappeared from the titles board and became invalid. Which made the title El presidente become active. So i became el presidente without holding elections or actually selecting anything. Unfortunately this caused most of our laws to become invalid as the previous title was now invalid. It may have been due to two separate government buildings existing and passing laws simultaneously. I will take some screenshots of referenced laws shortly.

**Comment by smokeykiller51** [comment]
![image](https://user-images.githubusercontent.com/70777108/92309209-6b06f500-ef58-11ea-89d0-205f3c693a3b.png)
This one happened more recently because a law referencing the title became invalid which made the title invalid. Luckily this time i had added an amendment to my player name rather than the title.

**Comment by smokeykiller51** [comment]
It also removed said titles completely from my table.

**Comment by beastyn** [comment]
I've lost a law with 34 attempts.
![bandicam 2020-09-07 12-02-41-952](https://user-images.githubusercontent.com/27898520/92357996-2e76fd00-f102-11ea-9fbf-d69fed5cfbf0.jpg)
.

**Comment by smokeykiller51** [comment]
![image](https://user-images.githubusercontent.com/70777108/92409771-9889a500-f0f6-11ea-9ec0-f68975b03df2.png)
So logged into Panda today 8:30 Mountain time -7 azmountain time. I noticed that a now invalid title which had a wage set on it is now paying out but not at the rate that was originally set. The Title is still not at the table which it was originally made on.

**Comment by beastyn** [comment]
@smokeykiller51 thank you for the attention. It can be worse with the old servers as objects were created with different initialization logic.

**Comment by SlayksWood** [comment]
> Defined some districts using the Zoning Office (District Map 1), used /civics vote & /fin to finalize and enable a new district map.
> Then upon editing the District Map 1 (became District Map 2) and using /civics vote & /fin, I got an error in the map viewer stating that Disctricts Map 1 was missing.
> Trying to update Districts Map 2 to a new Districts Map (probably 3), the entire entry is gone from the Zoning Office.
> (see screenshots for more information).
> Cannot update any existing districts because there is no reference anymore.
> 
> (edit: I used the clearoldelections command after defining District Map 2, because there was a lot of clutter from old wrongfully created elections)
> 
> ![scrn1](https://user-images.githubusercontent.com/38540031/93671761-0f834e00-faa6-11ea-8880-4eb8f0124b3b.png)
> 
> ![scrn2](https://user-images.githubusercontent.com/38540031/93671762-1316d500-faa6-11ea-9041-f677404d1e61.png)

**Comment by SlayksWood** [comment]
> We created a title (Legal Minister) which was approved by election, but since it was using an unapproved voting system, we had to have a second election to actually activate it by switching to another voting method.
> 
> When the first election was done, a constitutional amendment election was made to grant this title the powers to draft laws at the same time as the election to activate it.
> 
> After both elections were done, the elected title disappeared from government office, and the amendment disappeared from the amendments table, but are both still in effect, unrevisable, although they can be viewed in the government (L) tab.
> 
> See screenshots:
> 
> Government office missing title:
> ![image](https://user-images.githubusercontent.com/20558591/93798712-06c18200-fc36-11ea-9a6b-44d6613c726f.png)
> 
> Amendment missing:
> ![image](https://user-images.githubusercontent.com/20558591/93798803-28226e00-fc36-11ea-8b9e-b280e934d232.png)
> 
> Amendment present in the government tab:
> ![image](https://user-images.githubusercontent.com/20558591/93798902-4d16e100-fc36-11ea-9a53-99f5be189dd1.png)
> 
> Title present in the government tab:
> ![image](https://user-images.githubusercontent.com/20558591/93798983-6a4baf80-fc36-11ea-8649-c2a8812f33c3.png)

> @nichlas0010 Thank you for the report. We need some additional information.
> 
> 1. Do you use any admin commands?
> 2. Can you provide log files when it happened? or will happen next time.
>    Instruction is here https://github.com/StrangeLoopGames/EcoIssues/wiki/Getting-the-Eco-game-log-file
> 3. How often does this happen to you?
>    We have issue with this problem #16601, but we can't reproduce it(


> 1. No admin commands were used AFAIK, and the host said that all legal configs are default templates.
> 2. I don't have the log files from when it happened, sorry.
> 3. Only happened this once, all other titles and amendments work fine. This is my first time playing the game so I haven't had the bug before, but the fellas I play with say it hasn't happened to them before either
> 
> The game does seem to know the title's not proper:
> ![image](https://user-images.githubusercontent.com/20558591/93890540-59587800-fce2-11ea-9987-4dfc21bcf054.png)
> Restarting the server fixed the amendment but the title is still absent from the government

**Comment by IcedForge** [comment]
My save file from Develop-61 build it is currently 100% reproducing the issue and it is with
Election process ( when a object is passed via voting ) and using /fin command to pass ANY civics object.
It works fine when using Executor permissions ( which bypasses elections ).

Add Law 1(can be any civics object) to a table, pass it then revise it (is now Law 2) and push for vote and it will remove Law 2
from the table, and only show Law 1 <removed> in the draft slot.
[Storage.zip](https://github.com/StrangeLoopGames/EcoIssues/files/5267956/Storage.zip)
@johnkslg

**Comment by IcedForge** [comment]
Made a new issue with a summary and link to this post to clean it up and keep it a bit managable ( on johns request ).
